distributed/scheduler.py

```diff
@@ -2321,6 +2326,25 @@ def transition_processing_erred(
             else:
                 failing_ts = ts.exception_blame  # type: ignore
 
+            if exception_text and traceback_text:
+                self.erred_tasks.appendleft(ts)
+                ex = exception_text.split("(", 1)[0]
+                # Hash based on group name and exception type name, so that
+                # will be considered the unit of deduplication.
+                if (ts.group_key, ex) not in self.group_exceptions:
+                    # Cap the number of unique exception types we are tracking
+                    if len(self.group_exceptions) > 100:
+                        self.group_exceptions.popitem(last=False)
+                    self.group_exceptions[(ts.group_key, ex)] = {
+                        "exception": exception_text,
+                        "traceback": traceback_text,
+                        "workers": set(),
+                        "count": 0,
+                    }
+                self.group_exceptions[(ts.group_key, ex)]["workers"].add(w or worker)
+                self.group_exceptions[(ts.group_key, ex)]["count"] += 1
+                self.group_exceptions[(ts.group_key, ex)]["last_seen"] = time()
+
```

Comment:
My goal here was actually to do a level of deduplication, effectively trying to hash based on `(group, exception_type)`, so that many similar-but-not-identical exceptions are consolidated into a single row.

> Perhaps we could wander through all task groups, find any with erred tasks, and then use those?

Yes, I can refactor to do something like that. One thing that I did find helpful was keeping errored tasks in a `deque` to force them to be in scope a bit longer. Otherwise it was too easy to loose task state for what is presumably a forensic exercise. So I might stick up for that data structure with some history a bit more than this deduplicated-dictionary.

Reply:
Okay, I've removed (most of) the state, and it now has a lighter footprint on the scheduler

Reply:
I think what we've found useful to have is the exception and the traceback on the exception for investigating what went wrong, but I agree that I wouldn't want multiple states for this. 

For the future (if this helps guide your design decisions) I can imagine: 

1. wanting to retry these tasks via the dashboard
2. Figuring out what input mapped to each failed task (in particular for large client.map calls). 

In our experience tasks generally fail:

- All with exactly the same exception/traceback (example: something in the submitted function is broken)
- All with the same exception/mostly the same traceback  (example: some service is just down)
- Singleish task failures (some specific input is bad)

Reply:
Thanks for the insights @gerrymanoim! A few thoughts inline:

> For the future (if this helps guide your design decisions) I can imagine:
> 
>     1. wanting to retry these tasks via the dashboard

This would be really cool, though we'd probably want to proceed carefully, since adding things which actually affect scheduler state to the dashboard brings in some serious security considerations (cf #6407). I don't think anything in this design would prevent us from doing this, however, should we figure out good ways of handling the security concerns.

>     2. Figuring out what input mapped to each failed task (in particular for large client.map calls).

I also added the exception information to the individual task page, so as a user one can go from `Exceptions -> individual task -> dependencies` all via links. Does that work for the type of investigations you are thinking about?

> 
>     * All with exactly the same exception/traceback (example: something in the submitted function is broken)
> 
>     * All with the same exception/mostly the same traceback  (example: some service is just down)
> 
>     * Singleish task failures (some specific input is bad)

An earlier version of this did some light attempts at deduplicating similar tasks (e.g., they are all the same, but the chunk/partition index is different, so the text is strictly different). I have some thoughts about how that might be done, but for the time being I've removed that logic in favor of simplicity. If huge numbers of similar tasks winds up being a problem with this approach, I'd certainly be interested in hearing about it!

Reply:
> since adding things which actually affect scheduler state to the dashboard brings in some serious security considerations 

Definitely a fair point. Maybe something that's better as a (future) jupyter widget instead of on the dashboard? I can imagine a whole UI wrapper for `as_completed` that gives you progress/tracebacks/restarts (just thinking out loud here). 

> Exceptions -> individual task -> dependencies

Does dependencies mean upstream nodes in the graph? Or the actual input (if any) a dask was passed? i.e. if I do

```python
def fail_if_two(x):
    if x == 2:
         raise RuntimeError("It was two!")

numbers_to_run = list(range(1000))
futures = client.map(fail_if_two, numbers_to_run)
```
will I be able to recover `2` via the dashboard? 


> If huge numbers of similar tasks winds up being a problem with this approach, I'd certainly be interested in hearing about it!

👍 will report back - thanks!

Reply:
We can consider dashboard pages that affect state.  Arguably the Dask Dashboard should be treated more securely than it currently is by most systems.  We'd want to iterate with a broader set of stakeholders here first, but I'm open to it.

> Does dependencies mean upstream nodes in the graph? Or the actual input (if any) a dask was passed? i.e. if I do

By dependencies I think that @ian-r-rose mostly means upstream nodes, and not inputs.  Inputs might be tricky currently because they're currently stored in serialized form in the scheduler (and the scheduler isn't supposed to unpack them).  This will change if we merge in https://github.com/dask/distributed/pull/6028 (on hold until stability issues stabilize (which it seems like maybe they have?)).

Reply:
> By dependencies I think that @ian-r-rose mostly means upstream nodes, and not inputs

That's correct, the upstream task nodes are visible, but their deserialized inputs are not. For your example, if I edit it to have some upstream tasks:
```python
def fail_if_two(x):
    if x == 2:
         raise RuntimeError("It was two!")

def inc(x):
    return x + 1

numbers_to_run = list(range(1000))

futures_a = client.map(inc, numbers_to_run)
futures_b = client.map(fail_if_two, futures_a)
```

I will be able to see that the exception has an `inc` task as a dependency, but not what the actual input value was.

Reply:
Understood - thanks for the clarifications.